contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,851 @@
+that it has created from the PodTemplate that corresponds to the target 
+revision of the StatefulSet.
+1. The controller will then persist the StatefulSetStatus make it durable and 
+communicate it to observers.
+
+## API Server
+The API Server will perform validation for StatefulSet updates and ensure that 
+a StatefulSet's `.Spec.TemplateRevision` is a generator for a strictly 
+monotonically increasing sequence.
+
+### StatefulSet Validation
+As is currently implemented, the API Server will not allow mutation to any 
+fields of the StatefulSet object other than `.Spec.Replicas` and 
+`.Spec.Template.Containers`. This design imposes the following, additional 
+constraints.
+
+1. The `.Spec.RevisionHistoryDepty` must be greater than or equal to `2`. 
+1. The `.Spec.PositionOrdinal` must be in the sequence `[0,.Spec.Replicas)`. 
+
+### TemplateRevision Maintenance
+It will be the responsibility of the API Server to enforce that updates to 
+StatefulSet's `.Spec.Template` atomically increment the 
+`.Spec.TemplateRevision` counter. There is no need for the value to be 
+strictly sequential, but it must be strictly, monotonically increasing.
+As validation will not allow mutation to any field other than the 
+`.Spec.Template.Containers` field, the API Server need not track all fields of 
+StatefulSet's `.Spec` for modifications, but it must trigger an update to the 
+revision when the current and previous `.Spec.Template` versions fail a test for
+deep semantic equality.
+
+## Kubectl
+Kubectl will  use the `rollout` command to control and provide the status of 
+StatefulSet updates.
+
+ - `kubectl rollout status statefulset <StatefulSet-Name>`: displays the status 
+ of a StatefulSet update.
+ - `kubectl rollout undo statefulset <StatefulSet-Name>`: triggers a rollback 
+ of the current update.
+ - `kubectl rollout history statefulset <StatefulSet-Name>`: displays a the 
+ StatefulSets revision history.
+
+## Usage
+This section demonstrates how the design functions in typical usage scenarios.
+
+### Initial Deployment
+Users can create a StatefulSet using `kubectl create`.
+
+Given the following manifest `web.yaml`
+
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.8
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+Users can use the following command to create the StatefulSet.
+
+```shell
+kubectl create -f web.yaml
```

Comment:
If you plan to use `apply` later, you should use `apply` for the initial creation as well: https://kubernetes.io/docs/user-guide/kubectl/v1.6/#apply